core/primitives/src/time.rs

```diff
@@ -1,155 +1,295 @@
+                if clock.is_mock {
+                    let x = clock.pop_utc();
+                    match x {
+                        Some(t) => t,
+                        None => {
+                            panic!("Mock clock run out of samples");
+                        }
+                    }
+                } else {
+                    chrono::Utc::now()
+                }
+            })
+        }
+
+        pub fn instant() -> Instant {
+            MockClockPerThread::with(|clock| {
+                if clock.is_mock {
+                    let x = clock.pop_instant();
+                    match x {
+                        Some(t) => t,
+                        None => {
+                            panic!("Mock clock run out of samples");
+                        }
+                    }
+                } else {
+                    Instant::now()
+                }
+            })
+        }
+
+        pub fn now() -> Time {
+            Time::from_utc(Clock::utc())
+        }
+
+        pub fn instant_call_count() -> u64 {
+            MockClockPerThread::with(|clock| clock.instant_call_count)
+        }
 
-impl Default for MockClockGuard {
-    fn default() -> Self {
-        Clock::set_mock();
-        Self {}
+        pub fn utc_call_count() -> u64 {
+            MockClockPerThread::with(|clock| clock.utc_call_count)
+        }
     }
-}
 
-impl Drop for MockClockGuard {
-    fn drop(&mut self) {
-        Clock::reset();
+    #[cfg(test)]
+    mod tests {
+        // TODO(#5345) Add tests for mock
+        use super::*;
+
+        #[test]
+        fn test_mock() {
+            Clock::set_mock();
+            let utc = Utc::now();
+
+            Clock::add_utc(utc);
+
+            assert_eq!(Clock::utc(), utc);
+        }
     }
 }
 
-impl Clock {
-    pub fn set_mock() {
-        MockClockPerThread::with(|clock| {
-            clock.is_mock = true;
-        });
-    }
-    pub fn reset() {
-        MockClockPerThread::with(|clock| {
-            clock.reset();
-        });
+mod time {
+    use crate::time::{Clock, Duration, Utc};
+    use borsh::{BorshDeserialize, BorshSerialize};
+    use chrono::DateTime;
+    use std::ops::{Add, Sub};
+    use std::time::SystemTime;
+
+    pub const TIME_UNIX_EPOCH: Time = Time::from_system_time(SystemTime::UNIX_EPOCH);
+
+    #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+    pub struct Time {
+        time: SystemTime,
     }
-    pub fn add_utc(mock_date: DateTime<chrono::Utc>) {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                clock.utc.push_back(mock_date);
-            } else {
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+
+    impl BorshSerialize for Time {
+        fn serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
+            let nanos = self.duration_since(&TIME_UNIX_EPOCH).as_nanos() as u64;
```

Comment:
thought: this is subtle. `duration_since` is saturating, so using it for serialization might break borsh consistency/canonicity guarantees and give us two times which serialize to the same bytes. However, because we use UNIX_EPOCH which is zero, we actually guaranteed to never hit overflow behavior.

Reply:
@matklad I would like to implement `to_unix_timestampt_nanos` without using `duration_since`. Any suggestions?
```
        pub fn to_unix_timestampt_nanos(&self) -> u64 {
            self.duration_since(&Time::UNIX_EPOCH).as_nanos() as u64
        }
```